src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -725,27 +736,480 @@ mod commutation_tests {
     // get UB unless they are both reads.
     fn protected_enforces_noalias() {
-        for rel1 in AccessRelatedness::all() {
-            for rel2 in AccessRelatedness::all() {
-                if rel1.is_foreign() == rel2.is_foreign() {
-                    // We want to check pairs of accesses where one is foreign and one is not.
-                    continue;
+        for [rel1, rel2] in <[AccessRelatedness; 2]>::exhaustive() {
+            // We want to check pairs of accesses where one is foreign and one is not.
+            precondition!(rel1.is_foreign() != rel2.is_foreign());
+            for [kind1, kind2] in <[AccessKind; 2]>::exhaustive() {
+                for mut state in LocationState::exhaustive() {
+                    let protected = true;
+                    precondition!(state.perform_access(kind1, rel1, protected).is_ok());
+                    precondition!(state.perform_access(kind2, rel2, protected).is_ok());
+                    // If these were both allowed, it must have been two reads.
+                    assert!(
+                        kind1 == AccessKind::Read && kind2 == AccessKind::Read,
+                        "failed to enforce noalias between two accesses that are not both reads"
+                    );
+                }
+            }
+        }
+    }
+
+    // Possible pairs of `AccessRelatedness` that can occur
+    // to two tags that are mutually foreign.
+    // The two tags are by convention named `x` and `y` throughout these tests.
+    // This includes everything except `(StrictChildAccess, StrictChildAccess)`
+    // because no tag can be a child of both.
+    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+    enum AccessRel2 {
+        ChildX,
+        ChildY,
+        Other,
+    }
+
+    impl Exhaustive for AccessRel2 {
+        fn exhaustive() -> Box<dyn Iterator<Item = Self>> {
+            use AccessRel2 as Rel;
+            Box::new(vec![Rel::ChildX, Rel::ChildY, Rel::Other].into_iter())
+        }
+    }
+
+    #[derive(Debug, Clone, Copy)]
+    struct Access2 {
+        rel2: AccessRel2,
+        kind: AccessKind,
+    }
+
+    impl Exhaustive for Access2 {
+        fn exhaustive() -> Box<dyn Iterator<Item = Self>> {
+            Box::new(
+                <(AccessRel2, AccessKind)>::exhaustive().map(|(rel2, kind)| Access2 { rel2, kind }),
+            )
+        }
+    }
+
+    #[derive(Clone, PartialEq, Eq, Hash)]
+    /// Our testing utilties include the protector bound to the state
+    /// rather than the tag. There is little room for errors since we only
+    /// do examples on single-byte locations anyway.
+    struct LocStateProt {
+        state: LocationState,
+        prot: bool,
+    }
+
+    impl LocStateProt {
+        fn perform_access(&self, kind: AccessKind, rel: AccessRelatedness) -> Result<Self, ()> {
+            let mut state = self.state;
+            let _ = state.perform_access(kind, rel, self.prot).map_err(|_| ())?;
```

Comment:
```suggestion
            state.perform_access(kind, rel, self.prot).map_err(|_| ())?;
```
or is there a reason to explicitly use an `_` here?